### PR TITLE
User configurable TX button actions

### PR DESCRIPTION
**Pull request by pkendall64**
Adds the ability for a TX to have 2 user configurable buttons.
The actions of the button(s) can be configured in the web UI (shown below)
Screen Shot 2022-08-11 at 2 15 55 PM

A button can have up to 2 actions associated with it.
- Either one short-press and/or one long-press activation
- or two long-press activations

The actions that can be performed are:
- None (the button does nothing; useful if a user only wants a single action on the button)
- Increase Power (Increments the power output by one notch i.e. 25 -> 50mW, 50 -> 100mW etc)
- Go to VTX Band Menu (Jumps directly to the menu on the TFT/OLED display, if present)
- Go to VTX Channel Menu (Jumps directly to the menu on the TFT/OLED display, if present)
- Send VTX Settings (Sends the current VTX admin settings to the FC/Backpack)
- Start WiFi (Immediately starts WiFi mode)
- Enter Binding Mode (Tells the TX to start sending bind messages)

The framework of actions is extensible and more actions can be supported.

The Count column
- In Short-press mode, the count is the number of clicks (button presses) that must be done to execute the action.
- In Long-press mode, the count is the time the button must be held down before the action is executed, the time is in half-second increments.

### CODE REVIEW

**Comment by commanderguy3001** [comment]
are there any protections for the case that one has the "start wifi" or "reboot" option set, and accidentally hits it mid flight?

**Comment by deadbytefpv** [comment]
> are there any protections for the case that one has the "start wifi" or "reboot" option set, and accidentally hits it mid flight?

Arm state is being checked for buttons, joysticks.

**Comment by pkendall64** [comment]
> are there any protections for the case that one has the "start wifi" or "reboot" option set, and accidentally hits it mid flight?

I removed the reboot option as thats actually only for receivers and they are not configurable.
I have not put a check on the wifi option as the only check we could do is `CRSF::IsArmed` and if I add that Cap will rip me a new one!

**Comment by pkendall64** [comment]
> > are there any protections for the case that one has the "start wifi" or "reboot" option set, and accidentally hits it mid flight?
> 
> Arm state is being checked for buttons, joysticks.

Yes, thats right, all the buttons on the modules are protected already by the Armed state check in the driver for them.

**Comment by commanderguy3001** [comment]
ah ok, if all buttons are protected, even custom ones, then that should be good enough.

File: src/lib/CONFIG/config.h [hunk 1]
```diff
@@ -14,7 +14,7 @@
 #define TX_CONFIG_MAGIC     (0b01U << 30)
 #define RX_CONFIG_MAGIC     (0b10U << 30)
 
-#define TX_CONFIG_VERSION   7U
+#define TX_CONFIG_VERSION   8U
```
Comment (CapnBry) [inline src/lib/CONFIG/config.h:17]: Why does the config version need to change when the struct size stays the same and is already initialized to 0? This is why we allocated the bytes for the buttons in the config PR.

Comment (pkendall64) [reply src/lib/CONFIG/config.h:17]: It probably doesn't need to be updated. I'll check and remove if it's not required.

Comment (pkendall64) [reply src/lib/CONFIG/config.h]: Yup, not needed, removed!

**Comment by pkendall64** [comment]
Come to think of it, we should remove the "Increase Power" option because we've made it that the buttons can't be used when armed anyways 😆

**Comment by CapnBry** [comment]
Awesome that the config changes could be removed! I've scrolled through this but haven't had a chance to play with it yet. It's a big one, so I should have some time for it this week.

**Comment by pkendall64** [comment]
> Awesome that the config changes could be removed! I've scrolled through this but haven't had a chance to play with it yet. It's a big one, so I should have some time for it this week.

It's not thaaaat big, don't you remember the Unified one?

File: src/lib/BUTTON/devButton.h [hunk 2]
```diff
@@ -3,7 +3,41 @@
 #include "device.h"
 #include "config.h"
 
+#include <functional>
+
+typedef enum : uint8_t {
```
Comment (CapnBry) [inline src/lib/BUTTON/devButton.h:8]: `// Limited to 16 possible ACTIONs by config storage currently`

File: src/src/rx_main.cpp [hunk 3]
```diff
@@ -1361,6 +1367,20 @@ RF_PRE_INIT()
 }
 #endif
 
+void reboot()
```
Comment (CapnBry) [inline src/src/rx_main.cpp:1370]: This is not supposed to just reboot.
The 5 count long press of the RX is supposed to start the wifi (on ESP/ESP32), and 9 count is supposed to wipe the config. This functionality should be preserved. For sure with the config reset, that's the most important function the button has in any of our code anywhere currently. The wifi starting is also quite helpful so that would be nice to keep as well :/

**Review by CapnBry — CHANGES_REQUESTED** [review]
I haven't validated the functionality yet, this is just code changes. In addition, I think we need to stop using std::function as its use bloats code and data segment size quite a bit. After all it is a class and the storage associated with it and the effects of passing and assigning its value tends to invoke copy constructors all over. In this code replacing std::function with a function pointer saves 96 bytes of RAM on both ESP and STM32 platforms, as well as 188 bytes of code on STM32 or 240 bytes on ESP.
```cpp
diff --git a/src/lib/BUTTON/devButton.cpp b/src/lib/BUTTON/devButton.cpp
index edf2c7a7..0cfcf5c8 100644
--- a/src/lib/BUTTON/devButton.cpp
+++ b/src/lib/BUTTON/devButton.cpp
@@ -33,14 +33,10 @@ static constexpr struct {
 };
 #endif

-static std::function<void()> actions[ACTION_LAST] = { nullptr };

-const std::function<void()> *getButtonFunctions()
-{
-    return actions;
-}
+static ButtonAction_fn actions[ACTION_LAST] = { nullptr };

-void registerButtonFunction(action_e action, std::function<void()> function)
+void registerButtonFunction(action_e action, ButtonAction_fn function)
 {
     actions[action] = function;
 }
diff --git a/src/lib/BUTTON/devButton.h b/src/lib/BUTTON/devButton.h
index 15054579..851d4904 100644
--- a/src/lib/BUTTON/devButton.h
+++ b/src/lib/BUTTON/devButton.h
@@ -3,8 +3,6 @@
 #include "device.h"
 #include "config.h"

-#include <functional>
-
 typedef enum : uint8_t {
     ACTION_NONE,
     ACTION_INCREASE_POWER,
@@ -18,6 +16,8 @@ typedef enum : uint8_t {
     ACTION_LAST
 } action_e;

+typedef void (*ButtonAction_fn)();
+
 #if defined(GPIO_PIN_BUTTON)
     #if defined(TARGET_TX) || \
         (defined(TARGET_RX) && (defined(PLATFORM_ESP32) || defined(PLATFORM_ESP8266)))
@@ -35,9 +35,8 @@ typedef enum : uint8_t {
         action_e action;
     } action_t;

-    void registerButtonFunction(action_e action, std::function<void()> function);
-    const std::function<void()> *getButtonFunctions();
+    void registerButtonFunction(action_e action, ButtonAction_fn function);

 #else
-    inline void registerButtonFunction(uint8_t action, std::function<void()> function) {}
+    inline void registerButtonFunction(uint8_t action, ButtonAction_fn function) {}
 #endif
```
We don't use any of the extra functionality provided by the function class, we just misuse it as a way to pass a function pointer and should work to remove them. (I'm guilty of starting this I believe)

File: src/html/scan.js [hunk 4]
```diff
@@ -151,15 +151,13 @@ function initOptions() {
...
-      initNetwork();
+      updateOptions(data['options']);
+      updateConfig(data['config']);
+      scanTimer = setInterval(getNetworks, 2000);
     }
   };
-  xmlhttp.open('GET', '/options.json', true);
-  xmlhttp.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
+  xmlhttp.open('GET', '/config', true);
+  xmlhttp.setRequestHeader('Content-type', 'application/json');
```
[the remark this answers is not in the document]
Comment (pkendall64) [reply src/html/scan.js:160]: Yes the content-type should be removed as it's not required.

File: src/html/scan.js [hunk 5]
```diff
@@ -432,7 +457,120 @@ function updateOptions(data) {
...
+function updateButtons(data) {
+  buttonActions = data;
+  for (const [b, _v] of Object.entries(data)) {
+    for (const [p, v] of Object.entries(_v)) {
+      appendRow(parseInt(b), parseInt(p), v);
+    }
+  }
+}
+
+function validateSave() {
```
[the remark this answers is not in the document]
Comment (pkendall64) [reply src/html/scan.js:475]: Agree we need to pull it all apart and make it more cohesive, but not as part of this.
I'll think of better names for the functions though.

File: src/src/rx_main.cpp (hunk 3 above)
Comment (pkendall64) [reply src/src/rx_main.cpp:1370]: Add back the config reset and renamed function

File: src/lib/BUTTON/devButton.cpp [hunk 6]
```diff
@@ -100,6 +65,54 @@ static int start()
     {
         return DURATION_NEVER;
     }
+
+    if (GPIO_PIN_BUTTON != UNDEF_PIN)
+    {
+        button1.init(GPIO_PIN_BUTTON, GPIO_BUTTON_INVERTED);
+        button1.OnShortPress = [](){ handlePress(0, false, button1.getCount()); };
+        button1.OnLongPress = [](){ handlePress(0, true, button1.getCount()); };
```
Comment (CapnBry) [inline src/lib/BUTTON/devButton.cpp:73]: `button1.getLongCount()` not `getCount()`, same with button2

File: src/lib/BUTTON/devButton.cpp [hunk 7]
```diff
@@ -4,93 +4,58 @@
...
-    }
+    actions[action] = function;
 }
-#endif
 
-static void initialize()
+static void handlePress(uint8_t button, bool longPress, uint8_t count)
 {
-    if (GPIO_PIN_BUTTON != UNDEF_PIN)
+    std::list<action_t>::iterator it;
```
Comment (CapnBry) [inline src/lib/BUTTON/devButton.cpp:45]: iterator is unused

File: src/lib/BUTTON/devButton.cpp [hunk 8]
```diff
@@ -4,93 +4,58 @@
...
+    DBGLN("handle press");
```
Comment (CapnBry) [inline src/lib/BUTTON/devButton.cpp:46]: `DBGLN("handlePress(%u,%u,%u)", button, (uint8_t)longPress, count);` so this conveys some information, and switch the one in button.cpp to a `DBGVLN` so the same information isn't doubled up

File: src/lib/BUTTON/devButton.cpp [hunk 9]
```diff
@@ -100,6 +65,54 @@ static int start()
...
+#endif
+    }
+    if (GPIO_PIN_BUTTON2 != UNDEF_PIN)
+    {
+        button2.init(GPIO_PIN_BUTTON2, GPIO_BUTTON_INVERTED);
+        button2.OnShortPress = [](){ handlePress(1, false, button2.getCount()); };
+        button2.OnLongPress = [](){ handlePress(1, true, button2.getCount()); };
+#if defined(TARGET_TX)
+        const tx_button_color_t *button_actions = config.GetButtonActions(1);
+        if (button_actions->val.actions[0].action == ACTION_NONE && button_actions->val.actions[1].action == ACTION_NONE)
```
Comment (CapnBry) [inline src/lib/BUTTON/devButton.cpp:99]: I see what you're doing here, but we shouldn't have config.SetDefaults() spread all over our codebase. Just move the defaults into config.SetDefaults(). There is no need to increment the config version for this, because said hardware will either ship with this functioning already (and thus have the proper default built in), or will ship with no default if we don't get this out in time, which is easy enough for a user to change.

This code also will keep bringing this default back if the user hates the buttons and wants no action for them.

File: src/src/tx_main.cpp [hunk 10]
```diff
@@ -1006,6 +1023,11 @@ void setup()
     }
   }
 
+#if defined(HAS_BUTTON)
+  registerButtonFunction(ACTION_BIND, EnterBindingMode);
+  registerButtonFunction(ACTION_INCREASE_POWER, cyclePower);
```
Comment (CapnBry) [inline src/src/tx_main.cpp:1028]: `cyclePower` didn't work for me. It definitely immediately changed the power (log message SetPower X) but the OLED doesn't update and if I was connected, it immediately would reset back to the configured max power.

File: src/html/scan.js [hunk 11]
```diff
@@ -432,7 +455,120 @@ function updateOptions(data) {
...
+      <option value='true' ${v['is-long-press']===true ? 'selected' : ''}>Long press (hold)</option>
+    </select>
+    <label>Press</label>
+  </div>
+</td>
+<td>
+  <div class="mui-select" id="mui-long-${b}-${p}" style="display:${buttonActions[b][p]['is-long-press'] ? "block": "none"};">
+    <select id="select-long-${b}-${p}" onchange="changeCount(${b}, ${p}, this.value);">
+      <option value='' disabled hidden ${v['action']===0 ? 'selected' : ''}></option>
+      <option value='0' ${v['count']===0 ? 'selected' : ''}>for 0.5 seconds</option>
```
Comment (CapnBry) [inline src/html/scan.js:543]: The long presses list them as being in 0.5 second increments, but button.cpp works in full second increments, so if I set it to 2 seconds it took ~4 seconds to activate. Also note that the callback reports _repeat count_ not total count so the first long press is `getLongCount() == 0`. The lambda function might just add 1 to getLongCount to get actual seconds?

Half seconds is really weird as an option in my opinion too. Who would set something for 1 second vs 1.5 seconds? Maybe I don't have the reflexes to have things set for half second precision 😄

File: src/lib/WIFI/devWIFI.cpp [hunk 12]
```diff
@@ -281,59 +262,101 @@ static void HandleReset(AsyncWebServerRequest *request)
...
+      {
+        action.val.actions[pos].pressType = array[button][pos]["is-long-press"];
+        action.val.actions[pos].count = array[button][pos]["count"];
+        action.val.actions[pos].action = array[button][pos]["action"];
+      }
+      config.SetButtonActions(button, &action);
+    }
+    config.Commit();
+  }
+  request->send(204);
```
Comment (CapnBry) [inline src/lib/WIFI/devWIFI.cpp:345]: This sends 204 back according to devtools but there's no cuteAlert generated to say it was successful.

File: src/lib/WIFI/devWIFI.cpp [hunk 13]
```diff
@@ -281,59 +262,101 @@ static void HandleReset(AsyncWebServerRequest *request)
...
+  serializeJson(json, *response);
+  request->send(response);
+}
+
+#if defined(TARGET_TX)
+static void UpdateConfiguration(AsyncWebServerRequest *request, JsonVariant &json)
+{
+  if (json.containsKey("button-actions")) {
+    const JsonArray &array = json["button-actions"].as<JsonArray>();
+    for (int button=0 ; button<array.size() ; button++)
```
Comment (CapnBry) [inline src/lib/WIFI/devWIFI.cpp:332]: unsigned (generates warning since size is unsigned)

**Review by CapnBry — DISMISSED** [review]
Gettin' close, all the actions seem to work except the power one, which I think needs _something_. I didn't dig very deep.

File: src/html/scan.js (hunk 11 above)
Comment (pkendall64) [reply src/html/scan.js:543]: How is it every second when the MS_LONG in button.h is 500!?

File: src/lib/WIFI/devWIFI.cpp (hunk 12 above)
Comment (pkendall64) [reply src/lib/WIFI/devWIFI.cpp:345]: I'll fix it in the .js file.

File: src/lib/BUTTON/devButton.cpp (hunk 9 above)
Comment (pkendall64) [reply src/lib/BUTTON/devButton.cpp:99]: The problem that we have then is that there would be a circular dependency. config would need to include button to be able to put in the default values., but button depends on config so it can get the configuration.
PIO shits itself, and this has been the cause of so many problems for us.

File: src/src/tx_main.cpp (hunk 10 above)
Comment (pkendall64) [reply src/src/tx_main.cpp:1028]: I've changed the screen behaviour so that it will always display the current power, rather than configured power if non-dynamic and current power if dynamic. Also, the * next to the power is display if its dynamic of the current power is not the configured power.
The problem of it being reset back is an interaction between the g-sensor setting it back to the configured value if the handset goes from stationary to moving, and the dynamic power setting it back to configured on line 78 in `dynpower.cpp`, which is trying to adjust for the MinPower setting when you previously go too close to the RX!

File: src/lib/BUTTON/devButton.cpp (hunk 9 above)
Comment (pkendall64) [reply src/lib/BUTTON/devButton.cpp]: I moved the actions enum into `common.h` to avoid the circular dependency, and the default config is now in `config.cpp`, where it belongs.
Also updated the wifi code so you can pass /reset?config to reset the config to defaults as well.

File: src/html/scan.js (hunk 11 above)
Comment (pkendall64) [reply src/html/scan.js:543]: Maybe your reactions are faster than you think 😄

**Comment by JyeSmith** [comment]
Looks ok after some bench testing.  Itll be good to get merged and more use/testing with master.

**Review by JyeSmith — APPROVED** [review]

**Review by FOG-Yamato — APPROVED** [review]

**Review by CapnBry — COMMENTED** [review]
All functions working, all my concerns addressed, along with color selection now. I didn't even know there was a built in color picker in HTML now, but it works great!
